sycl/test-e2e/README.md

```diff
@@ -379,3 +379,104 @@ Note: please avoid using `REQUIRES: TEMPORARY_DISABLED` for this purpose, it's
 a non-standard mechanism. Use `UNSUPPORTED: true` instead, we track
 `UNSUPPORTED` tests using the mechanism described above. Otherwise the test
 risks remaining untraceable.
+
+## Compiling and executing tests on separate systems
+
+The execution of e2e tests can be separated into compilation and execution 
+stages via the `test-mode` lit parameter. This allows us to reduce testing time 
+by compiling tests on more powerful systems and reusing the binaries on other
+machines. By default the `test-mode` parameter is set to `full`, indicating 
+that both stages will run. This parameter can be set to `build-only`, or 
+`run-only`, to only run the compilation stage, or the execution stage 
+respectively.
+
+The modes work as follow:
+### `--param test-mode=full`
+This is the default mode tests run in. Tests are marked as unsupported if no
+device on the machine can fulfill the `REQUIRES`/`UNSUPPORTED` statements. In
+this mode all `RUN:` lines are executed normally, and two extra features are
+added: the `build-and-run-mode` and `run-mode`.
+
+To make a test only run in `full` mode add a `REQUIRES: build-and-run-mode` line.
+
+### `--param test-mode=build-only`
+This mode can be used to compile all test binaries. To do this all `UNSUPPORTED` 
+and `REQUIRES` statements are ignored unless they contain `UNSUPPORTED: true` or 
+`REQUIRES: build-and-run-mode` (NOTE: currently a WIP to make this mode properly 
+react to features). All `RUN:` lines within a test are ran in this mode unless 
+they contain the following expansions: `%{run}`, `%{run-unfiltered-devices}`, 
+or `%if run-mode`.
+
+Currently the only triple supported for `build-only` mode is `spir64`
+
+### `--param test-mode=run-only`
+In this mode tests will not be compiled, they will only run. To do this only the 
+`RUN:` lines that contain `%{run}`, `%{run-unfiltered-devices}` or `%if run-mode` 
+are executed. Tests are marked as unsupported in the same manner as `full` mode.
+Since tests are not compiled in this mode, for any test to pass the test 
+binaries should already be in the `test_exec_root` directory, either by having 
+ran `full` or `build-only` modes previously on the system, or having 
+transferred the test binaries into that directory. The `run-mode` feature is 
+added when in this mode.
+
+### Resolving common Issues with separate compilation and execution:
+A number of extra considerations need to be taken to write tests that are able
+to be compiled and executed on separate machines.
+
+- Tests that build and execute multiple binaries need to be written such that 
+the output of each compilation has a different name. This way no files are
+overwritten, and all the necessary binaries can be transferred to the running
+system.
+
+- Two scenarios need to be considered for tests that expectedly fail:
+  - Tests that are expected to fail on compilation, and thus also during 
+  execution, need to be marked as `XFAIL` with a feature that is device 
```

Comment:
Can we do `XFAIL: build-only`?

Reply:
I think something in this direction would be a solid idea however there are some specifics that I'd be unclear about. Because these tests would fail on compilation in `full` testing mode as well, I think instead of using `build-only` it would be more appropriate to do something like `XFAIL: build-mode`. The other question is what the behavior should be on `run-only` mode. Surely they would fail there too due to not having a binary, but adding this feature in that mode I don't think makes sense. Perhaps we could add it only when checking XFAILS? not too sure, open to ideas.